src/Interpreters/PreparedSets.h

```diff
@@ -19,6 +20,37 @@ class Set;
 using SetPtr = std::shared_ptr<Set>;
 class InterpreterSelectWithUnionQuery;
 
+/// Represents a set in a query that might be referenced at analysis time and built later during execution.
+/// Also it can represent a constant set that is ready to use.
+/// At analysis stage the FutureSets are created but not necessarily filled. Then for non-constant sets there
+/// must be an explicit step to build them before they can be used.
+/// FutureSet objects can be stored in PreparedSets and are not intended to be used from multiple threads.
+class FutureSet final
```

Comment:
Why we can't use `SharedSet` instead of `FutureSet`? Does `FutureSet` only provides helper methods like `isValid` and so on?

Reply:
The difference between SharedSet and FutureSet is that the code that uses FutureSet is only executed in one thread and it must never wait for the internal future object (there must be an explicit step that builds the Set and sets the value before it this value is used), that's why all wait-related methods are hidden. And SharedSet is used between multiple threads and provides a way to wait for the result.